src/IdentityModel.AspNetCore.OAuth2Introspection/Infrastructure/AsyncLazy.cs

```diff
@@ -2,14 +2,41 @@
 // Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
 
 using System;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace IdentityModel.AspNetCore.OAuth2Introspection.Infrastructure
 {
-    internal class AsyncLazy<T> : Lazy<Task<T>>
+    internal sealed class AsyncLazy<T>
     {
-        public AsyncLazy(Func<Task<T>> taskFactory) :
-            base(() => Task.Factory.StartNew(taskFactory).Unwrap())
-        { }
+        private Lazy<Task<T>> _lazyTaskFactory;
+        private readonly Func<Task<T>> _taskFactory;
+        private readonly object _mutex = new object();
+
+        public AsyncLazy(Func<Task<T>> taskFactory)
+        {
+            _taskFactory = taskFactory;
+            _lazyTaskFactory = InitLazy(_taskFactory);
+        }
+ 
+        public Task<T> GetAsync()
+        {
+            if (!_lazyTaskFactory.IsValueCreated || !_lazyTaskFactory.Value.IsFaulted) return _lazyTaskFactory.Value;
+
+            lock (_mutex)
+            {
+                if (_lazyTaskFactory.IsValueCreated && _lazyTaskFactory.Value.IsFaulted)
+                {
+                    _lazyTaskFactory = InitLazy(_taskFactory);
+                }
+
+                return _lazyTaskFactory.Value;
+            }
+        }
+
+        private static Lazy<Task<T>> InitLazy(Func<Task<T>> taskFactory)
+        {
+            return new Lazy<Task<T>>(() => Task.Factory.StartNew(taskFactory).Unwrap(), LazyThreadSafetyMode.ExecutionAndPublication);
```

Comment:
I don;t see what value the Task.Factory.StartNew provides. The thread that gets created will be immediatelybe passed back to the threadpool when the the factory hits an async call internally

Reply:
I'm just using the same initialization that was previously used. 

Can @leastprivilege comment on why `StartNew` was used?

Reply:
I can't remember. Feel free to change it.